File: Examples/Others/ObjCExample/ObjCExample/CustomPlugins/ConsentPlugin/ConsentPlugin.m
//
//  ConsentPlugin.m
//  ObjCExampleApp
//
//  Created by Satheesh Kannan on 31/08/26.
//

#import "ConsentPlugin.h"

#pragma mark - ConsentPlugin

@interface ConsentPlugin()

@property(nonatomic, retain) RSSAnalytics *client;
@property(nonatomic, strong) id<ConsentCategoryProvider> provider;

@end

@implementation ConsentPlugin
@synthesize pluginType;

/**
 Initializes the plugin with the specified CMP adapter.

 @param provider The adapter supplying the current consent lists and change
                 notifications.
 @return An initialized instance of ConsentPlugin.
 */
- (instancetype)initWithProvider:(id<ConsentCategoryProvider>)provider
{
    self = [super init];
    if (self) {
        self.provider = provider;
    }
    return self;
}

/**
 Returns the plugin type for this plugin.

 @return RSSPluginTypeUtility, indicating this plugin never intercepts events —
         it only reacts to the CMP and calls setConsent:.
 */
- (RSSPluginType)pluginType {
    return RSSPluginTypeUtility;
}

/**
 Subscribes to the CMP and pushes its current choices to the SDK.

 @param analytics The analytics client instance to be used by the plugin.
 */
- (void)setup:(RSSAnalytics * _Nonnull)analytics {
    self.client = analytics;

    __weak typeof(self) weakSelf = self;
    self.provider.onConsentChanged = ^{
        [weakSelf pushCurrentConsent];
    };

    [self pushCurrentConsent];
}

/**
 Stops listening to the CMP.

 Called when the plugin is removed from the analytics client.
 */
- (void)teardown {
    self.provider.onConsentChanged = nil;
}

/**
 Hands the CMP's current choices to the SDK.

 The supplied lists fully replace the current consent state; omitted values
 clear the corresponding list.
 */
- (void)pushCurrentConsent {
    RSSConsentManagementOptions *options = [[RSSConsentManagementOptions alloc]
                                            initWithAllowedConsentIds:self.provider.allowedConsentIds
                                                     deniedConsentIds:self.provider.deniedConsentIds];
    [self.client setConsent:options];
}

@end

File: Examples/Main/SwiftUIExample/SwiftUIExample/CustomPlugins/ConsentPlugin.swift
//
//  ConsentPlugin.swift
//  SwiftUIExampleApp
//
//  Created by Satheesh Kannan on 31/08/26.
//

import Foundation
import RudderStackAnalytics

// MARK: - ConsentPlugin
/**
 A sample pattern for bridging a Consent Management Platform into the SDK. This is example
 code, not SDK API — copy it into your project and adapt it to your CMP.
 
 The plugin never modifies the event context. The SDK owns `context.consentManagement` and
 stamps it from the state recorded by `setConsent(_:)`; a plugin writing that key is
 overwritten by the SDK and logs a warning.
 
 ## Usage:
 ```swift
 let plugin = ConsentPlugin(provider: myCmpAdapter)
 analytics.add(plugin: plugin)
 Adding the plugin pushes whatever the CMP already knows, then keeps the SDK in sync as the
 user changes their choices.
 */

final class ConsentPlugin: Plugin {
    /* Never intercepts events — it only reacts to the CMP and calls setConsent(_:). */
    var pluginType: PluginType = .utility
    
    /** The analytics client instance, set during setup. */
    var analytics: Analytics?
    
    private let provider: ConsentCategoryProvider
    
    /** Creates a new ConsentPlugin backed by the given CMP adapter. */
    init(provider: ConsentCategoryProvider) {
        self.provider = provider
    }
    
    /** Subscribes to the CMP and pushes its current choices to the SDK. */
    func setup(analytics: Analytics) {
        self.analytics = analytics
        self.provider.onConsentChanged = { [weak self] in
            self?.pushCurrentConsent()
        }
        self.pushCurrentConsent()
    }
    
    /** Stops listening to the CMP. */
    func teardown() {
        self.provider.onConsentChanged = nil
    }
    
    /**
     Hands the CMP's current choices to the SDK. The new lists fully replace the previous
     consent state and apply from the next event onward.
     */
    func pushCurrentConsent() {
        let options = ConsentManagementOptions(
            allowedConsentIds: self.provider.allowedConsentIds,
            deniedConsentIds: self.provider.deniedConsentIds
        )
        self.analytics?.setConsent(options)
    }
}

// MARK: - ConsentCategoryProvider
/**
 The slice of a Consent Management Platform that `ConsentPlugin` depends on.
 
 Adapt this to whichever CMP the app uses — the plugin needs only the two ID lists and a
 notification for when the user changes their choices.
 */
protocol ConsentCategoryProvider: AnyObject {
    /** Consent category IDs the user has granted. */
    var allowedConsentIds: [String] { get }
    
    /** Consent category IDs the user has denied. */
    var deniedConsentIds: [String] { get }
    
    /** Invoked by the CMP whenever the user's choices change. */
    var onConsentChanged: (() -> Void)? { get set }
}
